src/EditorFeatures/Core/Implementation/Structure/StructureTag.cs

```diff
@@ -0,0 +1,87 @@
+        {
+            Snapshot = snapshot;
+            OutliningSpan = blockSpan.TextSpan.ToSpan();
+            Type = ConvertType(blockSpan.Type);
+            IsCollapsible = blockSpan.IsCollapsible;
+            IsDefaultCollapsed = blockSpan.IsDefaultCollapsed;
+            IsImplementation = blockSpan.AutoCollapse;
+
+            // The HeaderSpan is what is used for drawing the guidelines and also what is shown if
+            // you mouse over a guideline. We will use the text from the hint start to the collapsing
+            // start; in the case this spans mutiple lines the editor will clip it for us and suffix an
+            // ellipsis at the end.
+            var headerSpan = Span.FromBounds(blockSpan.HintSpan.Start, blockSpan.TextSpan.Start);
+
+            if (headerSpan.Length > 0)
+            {
+                HeaderSpan = headerSpan;
+            }
+
+            CollapsedText = blockSpan.BannerText;
+            CollapsedHintFormSpan = blockSpan.HintSpan.ToSpan();
+            _tagProvider = tagProvider;
+        }
+
+        public ITextSnapshot Snapshot { get; }
+        public Span? OutliningSpan { get; }
+        public Span? HeaderSpan { get; }
+        public Span? GuideLineSpan => null;
+        public int? GuideLineHorizontalAnchorPoint => null;
+        public string Type { get; }
+        public bool IsCollapsible { get; }
+        public bool IsDefaultCollapsed { get; }
+        public bool IsImplementation { get; }
+
+        public readonly string CollapsedText;
+
+        /// <summary>
+        /// The contents of the buffer to show if we mouse over the collapsed indicator.
+        /// </summary>
+        public readonly Span CollapsedHintFormSpan;
+
+        public object? GetCollapsedForm()
+        {
+            return CollapsedText;
+        }
+
+        public object? GetCollapsedHintForm()
+        {
+            return _tagProvider.GetCollapsedHintForm(this);
+        }
+
+        private static string ConvertType(string type)
+        {
+            return type switch
+            {
+                BlockTypes.Conditional => PredefinedStructureTagTypes.Conditional,
+                BlockTypes.Comment => PredefinedStructureTagTypes.Comment,
+                BlockTypes.Expression => PredefinedStructureTagTypes.Expression,
+                BlockTypes.Imports => PredefinedStructureTagTypes.Imports,
+                BlockTypes.Loop => PredefinedStructureTagTypes.Loop,
+                BlockTypes.Member => PredefinedStructureTagTypes.Member,
+                BlockTypes.Namespace => PredefinedStructureTagTypes.Namespace,
+                BlockTypes.PreprocessorRegion => PredefinedStructureTagTypes.PreprocessorRegion,
+                BlockTypes.Statement => PredefinedStructureTagTypes.Statement,
+                BlockTypes.Type => PredefinedStructureTagTypes.Type,
+                _ => PredefinedStructureTagTypes.Structural
```

Comment:
assert here?

Reply:
Not really, this is the standard type (I think) and also used for other kinds too like region blocks.

@gundermanc: what does the editor even use this for?

Reply:
This API is adapted from the Productivity Power Tools extension, which used this to map the guides to different colors. For accessibility reasons, we haven't added this yet, but it's on the backlog as a potential improvement.

In the nearer term, there are some potential code navigation and editing features, like 'Move Code up/down/left/right' which have been prototyped using these.

If it's not too much work to implement this, it'd be great to do it as spec-ed.

Reply:
@gundermanc: there's a spec?

Reply:
Turn of phrase :) ... I mean to implement it by mapping to the closest [PredefinedStructureTagType](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.text.adornments.predefinedstructuretagtypes?view=visualstudiosdk-2019).

FWIW:

- There is a spec Word doc floating about from 2017 I can probably dig up, if desired.
- There is pretty decent documentation in [StructureTag](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.text.tagging.structuretag.-ctor?view=visualstudiosdk-2019#remarks) and its members.

Reply:
The documentation actually was good for the API, and right now yes we are mapping to the closest tag type; the unclear bit is what that was actually being used for.

Reply:
So I think short answer is this probably does what you want @gundermanc.

Reply:
Yep, LGTM